### DISCUSSION TITLE
Alembic Migration Checker: The Essential CI/CD Toolkit - GitHub Action

### DISCUSSION THREAD

**Discussion by DevGlitch** [sqlalchemy/alembic#1427] [Ideas]
👋 Hello Alembic users,

I've recently developed a GitHub Action, [**Alembic Migration Checker**](https://github.com/marketplace/actions/alembic-migration-checker), designed to enhance the reliability of database schema management within CI/CD pipelines. This action automates the comparison between the latest migration script's Alembic version and the current database state, proactively alerting to any discrepancies. It's crafted to be an essential tool for preventing migration conflicts and ensuring smooth deployments, with support for multi-pending migrations.

## Features:

- **Broad Database Support**: Works with PostgreSQL, MySQL, and SQLite.
- **Multi-Pending Migrations**: Seamlessly handles multiple pending migrations.
- **Detailed Exit Statuses**: Offers clear insights into your database schema's alignment with migration scripts.
- **CI/CD Integration**: Designed for easy integration into existing workflows.

## How It Works:

By integrating this action into your workflows, you can automate checks for pending migrations, ensuring your database schema is always up to date with your Alembic migrations directory.

## Recommended Use Cases:

- **Pull Requests**: Validate new migrations against the existing database schema to avoid conflicts and simplify code review.
- **Deployment**: Workflows: Ensure the database schema aligns with the repository's migration scripts before deployment.

<br>
This action is not just a tool but a step towards more reliable and manageable database migrations for anyone using Alembic in their development pipeline. I warmly invite the community to try it out, share your feedback, and contribute to its development. Whether you have ideas for new features, improvements, or questions, I'm all ears.
<br><br>

![Check it out on GitHub Marketplace](https://github.com/marketplace/actions/alembic-migration-checker) and view the ![repo for more details](https://github.com/DevGlitch/alembic-migration-checker).

Looking forward to hearing your thoughts and experiences!

Cheers,
DevGlitch

**Comment by matheudantas** [comment]
I was really shocked when an idea I had literally a week ago suddenly found someone who has a solution for it and coincidentally it's been 2 weeks since this post. Thanks universe :earth_americas:

**Reply by DevGlitch** [reply]
Perfect timing then! 😄Glad my project could be of help right when you needed it.

**Comment by cjw296** [comment]
Am I missing something or does this not do anything other than check that some downstream revision of `head` is in the `alembic_migrations` table?

How does it validate new migrations against the existing database schema?

How does it ensure the database schema aligns with the repository's migration scripts?

**Reply by DevGlitch** [reply]
The action confirms that the version of the database schema, as noted in the `alembic_version` table, matches the version of the pending and latest migration scripts. By focusing on ensuring version alignment, the action assumes that if the versions history match, the database schema is effectively aligned with the migration scripts' intentions. However, it does not conduct a detailed comparison between the SQL definitions in the migration scripts and the actual database schema, meaning it ensures procedural consistency rather than verifying the schema's structural integrity against the SQL changes.

I was already considering, for a potential version 2, to delve deeper by parsing SQL from migration scripts for direct schema comparison, ensuring enhanced alignment checks. However, developing such a feature involves significant complexity and requires dedicated time I currently don't have. I welcome contributions and am keen to work with the community to build out this robust version 2, addressing the nuanced verification needs that go beyond the current capabilities.

**Comment by mukul-mpac** [comment]
Some potential upgrades for future versions:

- support async requests
- provide functionality for `ssl` connections especially in the case of using `asyncpg` since it does not let you describe ssl mode in the connection string. Rather pass it to `create_engine` (something like [this](https://github.com/MagicStack/asyncpg/issues/737#issuecomment-1051771723))

Please correct me if any of these already exists and I have overlooked it, Thank you for your efforts. This is a great idea which can be super helpful!

**Reply by DevGlitch** [reply]
Thanks for your valuable suggestions!

- **Async Requests**: Since the action primarily performs a straightforward database query, adding async support might not be necessary and could complicate the code. The current synchronous approach is efficient for its purpose. Are there any specific use cases where you believe async support would be particularly beneficial?

- **SSL Connections**: Adding SSL support is an excellent idea for PostgreSQL connections. Similar configurations can also be applied for MySQL.

If you're interested in contributing these improvements, I'd encourage you to submit a pull request (PR). Your contributions would be greatly appreciated by the community, and it would be fantastic to collaborate on enhancing this tool!

Looking forward to your input!
